src/Apple/Provider.php

```diff
@@ -0,0 +1,204 @@
+class Provider extends AbstractProvider implements ProviderInterface
+{
+    /**
+     * Unique Provider Identifier.
+     */
+    const IDENTIFIER = 'APPLE';
+
+    private const URL = 'https://appleid.apple.com';
+
+    /**
+     * {@inheritdoc}
+     */
+    protected $scopes = [
+        'name',
+        'email',
+    ];
+
+    /**
+     * The separating character for the requested scopes.
+     *
+     * @var string
+     */
+    protected $scopeSeparator = ' ';
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function getAuthUrl($state)
+    {
+        return $this->buildAuthUrlFromBase(self::URL.'/auth/authorize', $state);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function getTokenUrl()
+    {
+        return self::URL.'/auth/token';
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function getCodeFields($state = null)
+    {
+        $fields = [
+            'client_id'     => $this->clientId,
+            'redirect_uri'  => $this->redirectUrl,
+            'scope'         => $this->formatScopes($this->getScopes(), $this->scopeSeparator),
+            'response_type' => 'code',
+            'response_mode' => 'form_post',
+        ];
+
+        if ($this->usesState()) {
+            $fields['state'] = $state;
+            $fields['nonce'] = Str::uuid().'-'.$state;
+        }
+
+        return array_merge($fields, $this->parameters);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getAccessTokenResponse($code)
+    {
+        $response = $this->getHttpClient()->post($this->getTokenUrl(), [
+            'headers'        => ['Authorization' => 'Basic '.base64_encode($this->clientId.':'.$this->clientSecret)],
+            'form_params'    => $this->getTokenFields($code),
+        ]);
+
+        return json_decode($response->getBody(), true);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function getTokenFields($code)
+    {
+        return array_merge(parent::getTokenFields($code), [
+            'grant_type' => 'authorization_code',
+        ]);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    protected function getUserByToken($token)
+    {
+        static::verify($token);
+        $claims = explode('.', $token)[1];
+
+        return json_decode(base64_decode($claims), true);
+    }
+
+    /**
+     * Verify apple jwt.
+     *
+     * @param $jwt
+     *
+     * @return object
+     *
+     * @see https://appleid.apple.com/auth/keys
+     */
+    public static function verify($jwt)
+    {
+        $signer = new Sha256();
+
+        $token = (new Parser())->parse((string) $jwt);
+
+        if ($token->getClaim('iss') !== self::URL) {
+            throw new InvalidStateException('Invalid Issuer', Response::HTTP_UNAUTHORIZED);
+        }
+        if ($token->isExpired()) {
+            throw new InvalidStateException('Token Expired', Response::HTTP_UNAUTHORIZED);
+        }
+
+        $data = json_decode(file_get_contents(self::URL.'/auth/keys'), true);
```

Comment:
Do these keys change?
At minimum, we should cache this

Reply:
Apple signs the JWT using the public keys. We are not sure if the keys would remain constant, Apple started with single public key and then started using two keys now. We can cache this for an hour assuming Apple will not refresh or change the public keys. If Apple changes the public keys the token will fail.

So should we cache this?

Reply:
Apple signs the JWT using the public keys. We are not sure if the keys would remain constant, Apple started with single public key and then started using two keys now. We can cache this for an hour assuming Apple will not refresh or change the public keys. If Apple changes the public keys the token will fail.

So should we cache this?

Reply:
Currently enabled cache for 5 minutes